ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilderForInsertOnly.java

```diff
@@ -0,0 +1,265 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.txn.compactor;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.hive.common.StatsSetupConst;
+import org.apache.hadoop.hive.metastore.api.*;
+import org.apache.hadoop.hive.ql.exec.DDLPlanUtils;
+import org.apache.hadoop.hive.ql.util.DirectionUtils;
+import org.apache.hive.common.util.HiveStringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Modifier;
+import java.util.*;
+
+/**
+ * Builds query strings that help with query-based compaction of insert-only tables.
+ */
+class CompactionQueryBuilderForInsertOnly extends CompactionQueryBuilder {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CompactionQueryBuilderForInsertOnly.class.getName());
+
+  /**
+   * Construct a CompactionQueryBuilderForInsertOnly with required params.
+   *
+   * @param compactionType major or minor or rebalance, e.g. CompactionType.MAJOR.
+   *                       Cannot be null.
+   *                       Rebalance compaction is not supported for insert-only tables.
+   * @throws IllegalArgumentException if compactionType is null or the compaction type is REBALANCE
+   */
+  CompactionQueryBuilderForInsertOnly(CompactionType compactionType) {
+    super(compactionType, true);
+    if (CompactionType.REBALANCE.equals(compactionType)) {
+      throw new IllegalArgumentException("Rebalance compaction can only be supported on full ACID tables.");
+    }
+  }
+
+  @Override
+  protected void buildSelectClauseForInsert(StringBuilder query) {
+    switch (compactionType) {
```

Comment:
The goal of this commit is to split switch like this into classes?
So I think this should be split into: 
CompactionQueryBuilderForMajor and CompactionQueryBuilderForMinor

Reply:
Yeah, but then we should make the switch in the CompactionQueryBuilderForMajor and CompactionQueryBuilderForMinor classes, to differentiate between insert-only and acid tables. I choose to keep the logic for the insert-only tables in one class. What I can do is split them up to CompactionQueryBuilderForInsertOnlyMajor and CompactionQueryBuilderForInsertOnlyMinor. I am not sure if it would make the design better, but I am absolutely open for discussing it. What do you think?

Reply:
I think this in CompactionQueryBuilderFactory class breaks a little bit of the logic/aim of this commit:

    if (insertOnly) {
      return new CompactionQueryBuilderForInsertOnly(compactionType);
    } else {
      switch (compactionType) {
      case MAJOR:
        return new CompactionQueryBuilderForMajor();
      case MINOR:
        return new CompactionQueryBuilderForMinor();
      case REBALANCE:
        return new CompactionQueryBuilderForRebalance();
      }
    }
    
    I would keep this part and propagate the insertOnly property like:
    
     switch (compactionType) {
      case MAJOR:
        return new CompactionQueryBuilderForMajor(insertOnly);
      case MINOR:
        return new CompactionQueryBuilderForMinor(insertOnly);
      case REBALANCE:
        return new CompactionQueryBuilderForRebalance(insertOnly);
        
        
        and split this switch cases between them accordingly.

Reply:
I see you point. But the insert-only compaction is kind of a special case. I would like to keep it separately from the ACID compactions. The original aim was to be able to change things in the compaction query generation logic more easily and by not breaking the other types of compactions. In my head, the separation was always like 'MAJOR', 'MINOR' and the compactions for the insert-only tables, because they are kind of specials. I would rather not put the insert-only logic into the acid compactions. I am ok to create separate files for the insert-only MAJOR and MINOR, so we can get rid of this switch case, but I wouldn't like to put the 'if (insertOnly)' choices into the MAJOR and MINOR compaction code.

Reply:
Ok, thank keep as it is in this change.